storage/client_test.go

```diff
@@ -196,14 +196,47 @@ func (m *multiTestContext) Stop() {
 // DistSender. This simple implementation will likely be incorrect in some
 // untested cases and is a temporary measure until DistSender is hooked up.
 func (m *multiTestContext) Send(ctx context.Context, call proto.Call) {
-	for _, ls := range m.senders {
-		ls.Send(ctx, call)
+	maxRetries := len(m.senders) * 2
+	// Iterate over all stores in the test context, looking for a suitable
+	// replica. In the case of a RangeKeyMismatchError, retry the request on the
+	// next store. In the case of a NotLeaderError, attempt the new leader next.
+	// Because NotLeaderErrors will adjust the index of our loop, a second
+	// "maxRetries" count is used to bound the number of retries.
+	for nextIdx, total := 0, 0; nextIdx < len(m.senders) && total < maxRetries; total++ {
+		call.Reply.Header().SetGoError(nil)
+		call.Args.Header().Replica = proto.Replica{}
+		m.senders[nextIdx].Send(ctx, call)
 		if err := call.Reply.Header().GoError(); err != nil {
-			// Try the next localSender if this localSender did not have the
-			// requested range.
-			if _, ok := err.(*proto.RangeKeyMismatchError); ok {
-				call.Reply.Header().SetGoError(nil)
+			switch err := err.(type) {
+			case *proto.RangeKeyMismatchError:
+				// Try the next localSender if this localSender did not have the
+				// requested range.
+				nextIdx++
 				continue
+			case *proto.NotLeaderError:
+				// localSender has the range, is *not* the Leader, but the
+				// Leader is not known; this can happen if the leader is removed
+				// from the group. Move the manual clock forward in an attempt to
+				// expire the lease.
+				if err.Leader == nil {
+					m.expireLeaderLeases()
+					continue
+				}
+				// If the leader IS known, retry the request against the leader.
+				foundLeader := false
+				for i, s := range m.stores {
+					if s.StoreID() == err.Leader.StoreID {
+						nextIdx = i
+						foundLeader = true
+						break
+					}
+				}
+				// If the current leader was known but wasn't found in our
+				// collection of stores, break out of the loop; this should
+				// never happen in a correctly configured test.
+				if foundLeader {
+					continue
```

Comment:
If this should never happen, make it a `panic` instead of `continue`. Continuing without either expiring leases or changing `nextIdx` seems wrong.